### PR TITLE
Cleanup cyclomatic complexity with function getScheduleType

### LINKED ISSUE #460: Cleanup cyclomatic complexity issue
**What would you like to be added**:
There are 4 functions in which cyclomatic complexity exceed 15:
- [x] 1. EnsureWork() with complexity 21: (@XiShanYongYe-Chang #473)
https://github.com/karmada-io/karmada/blob/dba61fa92e8d22fb2292f5e5cd50d188185d7a34/pkg/util/helper/binding.go#L154
- [x] 2. retainServiceFields() with complexity 20: (@XiShanYongYe-Chang #476 )
https://github.com/karmada-io/karmada/blob/dba61fa92e8d22fb2292f5e5cd50d188185d7a34/pkg/util/helper/binding.go#L154
- [x] 3. getScheduleType() with complexity 20: (@ESonata #685)
https://github.com/karmada-io/karmada/blob/dba61fa92e8d22fb2292f5e5cd50d188185d7a34/pkg/scheduler/scheduler.go#L323
- [x] 4. RunJoin() with complexity (@XiShanYongYe-Chang #636 )
https://github.com/karmada-io/karmada/blob/dba61fa92e8d22fb2292f5e5cd50d188185d7a34/pkg/karmadactl/join.go#L154

You can get the complexity on your side by following commands:
```
# go get github.com/fzipp/gocyclo/cmd/gocyclo@latest
# gocyclo -over 15 pkg
21 helper EnsureWork pkg/util/helper/binding.go:154:1
20 objectwatcher retainServiceFields pkg/util/objectwatcher/retain.go:61:1
20 scheduler (*Scheduler).getScheduleType pkg/scheduler/scheduler.go:323:1
18 karmadactl RunJoin pkg/karmadactl/join.go:154:1
```

I'd prefer to fix it by separated PRs(each PR fix one issue), forks can choose any issues.

**Why is this needed**:
A function with a higher cyclomatic complexity requires more test cases to cover all possible paths and is potentially harder to understand.

### PR DESCRIPTION
**Pull request by ESonata**
**What type of PR is this?**

/kind cleanup

**What this PR does**:

Cleanup cyclomatic complexity with function getScheduleType,make its  cyclomatic complexity below 15
>   https://github.com/karmada-io/karmada/blob/dba61fa92e8d22fb2292f5e5cd50d188185d7a34/pkg/scheduler/scheduler.go#L323

**Which issue(s) this PR fixes**:

Fixes #460 

**Does this PR introduce a user-facing change?**:
```release-note
     NONE
```

### CODE REVIEW

**Comment by XiShanYongYe-Chang** [comment]
@ESonata you can sign off refer to https://github.com/karmada-io/karmada/pull/685/checks?check_run_id=3472945151

by the way, please compress the commit.

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -882,3 +802,89 @@ func (s *Scheduler) scaleScheduleClusterResourceBinding(clusterResourceBinding *
...
+		boundCluster := tc.Name
+		for _, c := range clusters {
+			if c.Cluster().Name == boundCluster {
+				if meta.IsStatusConditionPresentAndEqual(c.Cluster().Status.Conditions, clusterv1alpha1.ClusterConditionReady, metav1.ConditionFalse) {
+					return FailoverSchedule
+				}
+			}
+		}
+	}
+	return nil
```
Comment (XiShanYongYe-Chang) [inline pkg/scheduler/scheduler.go:842]: If no match case above, here can be return `AvoidSchedule` according to the original logic.

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -882,3 +802,89 @@ func (s *Scheduler) scaleScheduleClusterResourceBinding(clusterResourceBinding *
...
```
Comment (XiShanYongYe-Chang) [inline pkg/scheduler/scheduler.go:889]: If no match case above, here can be return `AvoidSchedule` according to the original logic.

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -882,3 +802,89 @@ func (s *Scheduler) scaleScheduleClusterResourceBinding(clusterResourceBinding *
 	}
 	return nil
 }
+
+func (s *Scheduler) getTypeFromResourceBindings(ns, name string) interface{} {
```
Comment (XiShanYongYe-Chang) [inline pkg/scheduler/scheduler.go:806]: Return type can be `ScheduleType`. Same as follow.

**Comment by XiShanYongYe-Chang** [comment]
Nice job, compress the commit again.

**Comment by RainbowMango** [comment]
@XiShanYongYe-Chang Please take another look. 
Thanks @ESonata.

**Pull request merged**